Adding Reactive Pagination Support

trande4884 opened the pull request:
# Description

Adding pagination support to ReactiveRepository with integration tests. The response is a CosmosPagedFlux which can be used to access the return data via .byPage(). This only supports a page size and will not support offset and limit.

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

kushagraThapar commented on sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/ReactiveCosmosOperations.java:
```diff
@@ -82,6 +83,27 @@ Mono<CosmosContainerProperties> replaceContainerProperties(String containerName,
+     * @param <T>           type of domainType
+     * @return results as Page
+     */
+    <T> CosmosPagedFlux<T> findAll(Class<T> domainType, String containerName);
```
This API is find all items by partition key in a container. where is the partition key param here? :)
Also, we probably don't need the containerName - since that can be figured out by domainType.

kushagraThapar commented on sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/ReactiveCosmosOperations.java:
```diff
@@ -82,6 +83,27 @@ Mono<CosmosContainerProperties> replaceContainerProperties(String containerName,
+     * @param <T>           type class of domainType
+     * @return results as Page
+     */
+    <T> CosmosPagedFlux<T> paginationQuery(CosmosQuery query, Class<T> domainType, String containerName);
```
Do we need to expose containerName?
Also, this seems to be a very generic API that we are exposing, is there a customer request for this?

kushagraThapar replied:
For both these APIs, I wouldn't want to expose them directly in the repository if they are not required or requested. Since they are directly touching the public surface area. 
Also, what I understand from your feature is that now customers can create JPA queries with CosmosPagedFlux as return type. If that's correct, then these APIs can be added by customers themselves. 
We should mostly have only those APIs here which are generic enough and special at the same time to provide benefit to customer by some optimization, or fulfilling some necessity.

kushagraThapar commented on sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/repository/query/AbstractReactiveCosmosQuery.java:
```diff
@@ -61,12 +61,11 @@ public Object execute(Object[] parameters) {
         } else if (isPageQuery()) {
-            throw new IllegalArgumentException("Paged Query is not supported by reactive cosmos "
-                + "db");
+            return new ReactiveCosmosQueryExecution.PagedExecution(operations, accessor.getPageable());
```
`isPageQuery()` checks if the return type is of `Page.java` type, which in this case it is not. Since `CosmosPagedFlux` is not type of `Page.java`. How will it go in this else if condition?

kushagraThapar reviewed (commented):
@trande4884  I am wondering if the PR is complete. If so, we should update the title of the PR :)
I have added few comments which I think are blocking. Please take a look. thanks again!

trande4884 commented:
> @trande4884 I am wondering if the PR is complete. If so, we should update the title of the PR :) I have added few comments which I think are blocking. Please take a look. thanks again!

It is done, I thought I updated this the other day but must not have clicked save :)

trande4884 commented:
In looking into this more I do not think we can return an object which extends Flux.

I created a basic class:
`public class BasicFlux<T> extends Flux {
    @Override
    public void subscribe(CoreSubscriber coreSubscriber) {
    }
}`

When I try to return this as a type:
`BasicFlux<ReactiveTeacher> findAllByFirstName(String firstName);`

I get the following error:
`java.lang.ClassCastException: class reactor.core.publisher.FluxUsingWhen cannot be cast to class com.azure.spring.data.cosmos.common.BasicFlux (reactor.core.publisher.FluxUsingWhen and com.azure.spring.data.cosmos.common.BasicFlux are in unnamed module of loader 'app')

	at com.sun.proxy.$Proxy77.findAllByFirstName(Unknown Source)
	at com.azure.spring.data.cosmos.repository.integration.ReactiveTeacherRepositoryIT.testFindAllItemsOnePage(ReactiveTeacherRepositoryIT.java:83)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.springframework.test.context.junit4.statements.RunBeforeTestExecutionCallbacks.evaluate(RunBeforeTestExecutionCallbacks.java:74)
	at org.springframework.test.context.junit4.statements.RunAfterTestExecutionCallbacks.evaluate(RunAfterTestExecutionCallbacks.java:84)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.springframework.test.context.junit4.statements.RunBeforeTestMethodCallbacks.evaluate(RunBeforeTestMethodCallbacks.java:75)
	at org.springframework.test.context.junit4.statements.RunAfterTestMethodCallbacks.evaluate(RunAfterTestMethodCallbacks.java:86)
	at org.springframework.test.context.junit4.statements.SpringRepeat.evaluate(SpringRepeat.java:84)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:251)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:97)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70)
	at com.azure.spring.data.cosmos.AbstractIntegrationTestCollectionManager$1.evaluate(AbstractIntegrationTestCollectionManager.java:122)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:190)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:235)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:54)`
